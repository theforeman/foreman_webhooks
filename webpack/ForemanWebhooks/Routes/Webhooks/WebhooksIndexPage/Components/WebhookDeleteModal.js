import React from 'react';
import PropTypes from 'prop-types';

import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import ForemanModal from 'foremanReact/components/ForemanModal';
import { foremanUrl } from 'foremanReact/common/helpers';

import { WEBHOOK_DELETE_MODAL_ID } from '../../constants';

const WebhookDeleteModal = ({ toDelete, onSuccess }) => {
  const { id, name } = toDelete;

  return (
    <ForemanModal
      id={WEBHOOK_DELETE_MODAL_ID}
      title={__('Confirm Webhook Deletion')}
      backdrop="static"
      enforceFocus
      submitProps={{
        url: foremanUrl(`/api/v2/webhooks/${id}`),
        message: sprintf(__('Webhook %s was successfully deleted'), name),
        onSuccess,
        submitBtnProps: {
          bsStyle: 'danger',
          btnText: __('Delete'),
        },
      }}
    >
      {sprintf(__('You are about to delete %s. Are you sure?'), name)}
      <ForemanModal.Footer />
    </ForemanModal>
  );
};

WebhookDeleteModal.propTypes = {
  toDelete: PropTypes.object,
  onSuccess: PropTypes.func.isRequired,
};

WebhookDeleteModal.defaultProps = {
  toDelete: {},
};

export default WebhookDeleteModal;
