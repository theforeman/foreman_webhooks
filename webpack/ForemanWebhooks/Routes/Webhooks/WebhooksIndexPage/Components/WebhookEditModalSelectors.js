import { STATUS } from 'foremanReact/constants';
import { deepPropsToCamelCase } from 'foremanReact/common/helpers';
import {
  selectAPIStatus,
  selectAPIResponse,
} from 'foremanReact/redux/API/APISelectors';

import { WEBHOOK_API_REQUEST_KEY } from '../../constants';

export const selectWebhookValues = state =>
  deepPropsToCamelCase(selectAPIResponse(state, WEBHOOK_API_REQUEST_KEY)) || {};

export const selectWebhookTemplateId = state =>
  selectWebhookValues(state)?.webhookTemplate?.id || 0;

const selectWebhooksEditPageStatus = state =>
  selectAPIStatus(state, WEBHOOK_API_REQUEST_KEY);

export const selectIsLoading = state => {
  const status = selectWebhooksEditPageStatus(state);
  return !status || status === STATUS.PENDING;
};
