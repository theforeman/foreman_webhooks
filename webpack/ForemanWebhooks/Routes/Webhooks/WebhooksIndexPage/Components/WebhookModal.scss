.webhooks-modal {
  .modal-dialog {
    width: 50vw;
    min-width: 50vw;
  }
}
