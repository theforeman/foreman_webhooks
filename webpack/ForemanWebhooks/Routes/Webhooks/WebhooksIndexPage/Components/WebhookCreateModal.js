import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';

import { translate as __ } from 'foremanReact/common/I18n';
import { submitForm } from 'foremanReact/redux/actions/common/forms';
import { foremanUrl } from 'foremanReact/common/helpers';
import ForemanModal from 'foremanReact/components/ForemanModal';

import ConnectedWebhookForm from '../../Components/WebhookForm';

import { WEBHOOK_CREATE_MODAL_ID, WEBHOOKS_PATH } from '../../constants';

import './WebhookModal.scss';

const WebhookCreateModal = ({ onSuccess, onCancel }) => {
  const dispatch = useDispatch();

  const initialWebhookValues = {
    name: '',
    target_url: '',
    user: '',
    password: '',
    webhook_template_id: undefined,
    event: '',
    http_method: 'POST',
    http_content_type: 'application/json',
    enabled: true,
    verify_ssl: true,
    ssl_ca_certs: '',
    http_headers: '',
    proxy_authorization: false,
  };

  const handleSubmit = (values, actions) =>
    dispatch(
      submitForm({
        url: foremanUrl(`/api${WEBHOOKS_PATH}`),
        values: { ...values, controller: 'webhooks' },
        item: 'Webhook',
        message: __('Webhook was successfully created.'),
        successCallback: onSuccess,
        actions,
      })
    );

  return (
    <ForemanModal
      id={WEBHOOK_CREATE_MODAL_ID}
      title={__('Create Webhook')}
      backdrop="static"
      className="webhooks-modal"
    >
      <ForemanModal.Header />
      <ConnectedWebhookForm
        handleSubmit={handleSubmit}
        initialValues={initialWebhookValues}
        onCancel={onCancel}
      />
    </ForemanModal>
  );
};

WebhookCreateModal.propTypes = {
  onSuccess: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default WebhookCreateModal;
