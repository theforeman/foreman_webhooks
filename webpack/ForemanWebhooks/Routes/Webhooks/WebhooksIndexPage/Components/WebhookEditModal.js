import React, { useEffect, useState } from 'react';
import { Modal } from 'patternfly-react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { translate as __ } from 'foremanReact/common/I18n';
import ForemanModal from 'foremanReact/components/ForemanModal';
import Loading from 'foremanReact/components/Loading';
import { foremanUrl } from 'foremanReact/common/helpers';
import { submitForm } from 'foremanReact/redux/actions/common/forms';
import { get } from 'foremanReact/redux/API';

import ConnectedWebhookForm from '../../Components/WebhookForm';

import {
  WEBHOOKS_PATH,
  WEBHOOK_API_REQUEST_KEY,
  WEBHOOK_EDIT_MODAL_ID,
  WEBHOOKS_API_PLAIN_PATH,
} from '../../constants';

import {
  selectIsLoading,
  selectWebhookValues,
  selectWebhookTemplateId,
} from './WebhookEditModalSelectors';

import './WebhookModal.scss';

const WebhookEditModal = ({ toEdit, onSuccess, onCancel }) => {
  const dispatch = useDispatch();

  const [isPasswordDisabled, setIsPasswordDisabled] = useState(false);
  const id = toEdit;

  const isLoading = useSelector(selectIsLoading);
  const isPasswordSet = useSelector(selectWebhookValues).passwordSet;
  const initialWebhookValues = {
    id: useSelector(selectWebhookValues).id,
    name: useSelector(selectWebhookValues).name,
    target_url: useSelector(selectWebhookValues).targetUrl,
    user: useSelector(selectWebhookValues).user || '',
    password: '',
    http_method: useSelector(selectWebhookValues).httpMethod,
    http_content_type: useSelector(selectWebhookValues).httpContentType || '',
    webhook_template_id: useSelector(selectWebhookTemplateId),
    event: useSelector(selectWebhookValues).event,
    enabled: useSelector(selectWebhookValues).enabled,
    verify_ssl: useSelector(selectWebhookValues).verifySsl,
    ssl_ca_certs: useSelector(selectWebhookValues).sslCaCerts || '',
    http_headers: useSelector(selectWebhookValues).httpHeaders || '',
    proxy_authorization: useSelector(selectWebhookValues).proxyAuthorization,
  };

  useEffect(() => {
    setIsPasswordDisabled(isPasswordSet);
  }, [isPasswordSet]);

  const handleSubmit = (values, actions) => {
    if (isPasswordDisabled) {
      delete values.password;
    }
    dispatch(
      submitForm({
        url: foremanUrl(`/api${WEBHOOKS_PATH}/${id}`),
        values: { ...values, controller: 'webhooks' },
        item: 'Webhook',
        message: __('Webhook was successfully updated.'),
        method: 'put',
        successCallback: onSuccess,
        actions,
      })
    );
  };

  useEffect(() => {
    if (id) {
      dispatch(
        get({
          key: WEBHOOK_API_REQUEST_KEY,
          url: foremanUrl(`${WEBHOOKS_API_PLAIN_PATH}/${id}`),
        })
      );
    }
  }, [id, dispatch]);

  const onEditCancel = () => {
    if (isPasswordSet) setIsPasswordDisabled(true);
    onCancel();
  };

  return (
    <ForemanModal
      id={WEBHOOK_EDIT_MODAL_ID}
      backdrop="static"
      className="webhooks-modal"
    >
      <Modal.Header>
        <Modal.Title>
          {`${__('Edit')} ${initialWebhookValues.name}`}
        </Modal.Title>
      </Modal.Header>
      {isLoading ? (
        <Loading />
      ) : (
        <ConnectedWebhookForm
          handleSubmit={handleSubmit}
          initialValues={initialWebhookValues}
          onCancel={onEditCancel}
          isPasswordDisabled={isPasswordDisabled}
          setIsPasswordDisabled={setIsPasswordDisabled}
        />
      )}
    </ForemanModal>
  );
};

WebhookEditModal.propTypes = {
  onSuccess: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  toEdit: PropTypes.number,
};

WebhookEditModal.defaultProps = {
  toEdit: 0,
};

export default WebhookEditModal;
