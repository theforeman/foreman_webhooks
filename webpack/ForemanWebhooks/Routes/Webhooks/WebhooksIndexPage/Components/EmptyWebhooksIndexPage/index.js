import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import DefaultEmptyState from 'foremanReact/components/common/EmptyState';

import { useForemanModal } from 'foremanReact/components/ForemanModal/ForemanModalHooks';

import WebhookCreateModal from '../WebhookCreateModal';
import { WEBHOOK_CREATE_MODAL_ID } from '../../../constants';

const EmptyWebhooksIndexPage = ({
  search,
  reloadWithSearch,
  message: { type, text },
}) => {
  const {
    setModalOpen: setCreateModalOpen,
    setModalClosed: setCreateModalClosed,
  } = useForemanModal({
    id: WEBHOOK_CREATE_MODAL_ID,
  });

  return (
    <>
      <WebhookCreateModal
        onSuccess={() => {
          setCreateModalClosed();
          reloadWithSearch(search);
        }}
        onCancel={setCreateModalClosed}
      />
      <DefaultEmptyState
        icon={type === 'error' ? 'error-circle-o' : 'add-circle-o'}
        header={type === 'error' ? __('Error') : __('No Results')}
        description={text}
        documentation={null}
        action={{
          title: __('Create Webhook'),
          onClick: () => setCreateModalOpen(),
        }}
      />
    </>
  );
};

EmptyWebhooksIndexPage.propTypes = {
  message: PropTypes.shape({
    type: PropTypes.oneOf(['empty', 'error']),
    text: PropTypes.string,
  }),
  search: PropTypes.string,
  reloadWithSearch: PropTypes.func.isRequired,
};

EmptyWebhooksIndexPage.defaultProps = {
  message: {
    type: 'empty',
    text: __('Try to create a new Webhook'),
  },
  search: '',
};

export default EmptyWebhooksIndexPage;
