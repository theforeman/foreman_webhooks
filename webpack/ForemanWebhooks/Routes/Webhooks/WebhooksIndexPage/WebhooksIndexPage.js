import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import TableIndexPage from 'foremanReact/components/PF4/TableIndexPage/TableIndexPage';
import { translate as __ } from 'foremanReact/common/I18n';
import { useForemanModal } from 'foremanReact/components/ForemanModal/ForemanModalHooks';

import {
  WEBHOOKS_API_PATH,
  WEBHOOKS_API_REQUEST_KEY,
  WEBHOOK_CREATE_MODAL_ID,
} from '../constants';

import { selectSearch } from '../WebhooksPageSelectors';

import WebhooksTable from './Components/WebhooksTable';
import WebhookCreateModal from './Components/WebhookCreateModal';

import { reloadWithSearch, fetchAndPush } from '../WebhooksPageActions';

const WebhooksIndexPage = () => {
  const dispatch = useDispatch();

  const search = useSelector(selectSearch);

  const [toDelete, setToDelete] = useState({});
  const [toTest, setToTest] = useState({});
  const [toEdit, setToEdit] = useState(0);

  const {
    setModalOpen: setCreateModalOpen,
    setModalClosed: setCreateModalClosed,
  } = useForemanModal({
    id: WEBHOOK_CREATE_MODAL_ID,
  });

  return (
    <>
      <WebhookCreateModal
        onSuccess={() => {
          setCreateModalClosed();
          dispatch(reloadWithSearch(search));
        }}
        onCancel={setCreateModalClosed}
      />
      <TableIndexPage
        header={__('Webhooks')}
        controller="webhooks"
        apiUrl={WEBHOOKS_API_PATH}
        apiOptions={{ key: WEBHOOKS_API_REQUEST_KEY }}
        customCreateAction={() => setCreateModalOpen}
      >
        <WebhooksTable
          fetchAndPush={params => dispatch(fetchAndPush(params))}
          toDelete={toDelete}
          setToDelete={setToDelete}
          toEdit={toEdit}
          setToEdit={setToEdit}
          toTest={toTest}
          setToTest={setToTest}
          reloadWithSearch={query => dispatch(reloadWithSearch(query))}
        />
      </TableIndexPage>
    </>
  );
};

export default WebhooksIndexPage;
