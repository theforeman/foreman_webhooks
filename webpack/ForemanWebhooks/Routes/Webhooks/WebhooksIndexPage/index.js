import WebhooksIndexPage from './WebhooksIndexPage';

export default WebhooksIndexPage;
