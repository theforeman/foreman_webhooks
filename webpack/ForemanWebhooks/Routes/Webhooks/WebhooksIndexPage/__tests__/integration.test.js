import React from 'react';
import { IntegrationTestHelper } from '@theforeman/test';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import ConnectedWebhooksIndexPage from '../index';

import * as selectors from '../../WebhooksPageSelectors';
import * as editSelectors from '../Components/WebhookEditModalSelectors';
import { spySelector, spyEditSelector } from './WebhooksIndexPage.fixtures';

spySelector(selectors);
spyEditSelector(editSelectors);

describe('WebhooksIndexPage - Integration Test', () => {
  it('should flow', () => {
    const history = createMemoryHistory();
    history.push({ pathname: '/webhooks', search: '' });

    const integrationTestHelper = new IntegrationTestHelper();

    const component = integrationTestHelper.mount(
      <Router history={history}>
        <ConnectedWebhooksIndexPage history={history} />
      </Router>
    );
    integrationTestHelper.takeStoreAndLastActionSnapshot('rendered');

    expect(component.exists('WebhooksTable')).toEqual(true);
    expect(component.exists('WebhookCreateModal')).toEqual(true);

    integrationTestHelper.takeStoreAndLastActionSnapshot('initial state');
  });
});
