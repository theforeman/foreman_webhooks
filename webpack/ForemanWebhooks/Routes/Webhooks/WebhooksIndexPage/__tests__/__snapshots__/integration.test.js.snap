// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhooksIndexPage - Integration Test should flow: initial state 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "key": "WEBHOOK_EVENTS",
        "type": "get-some-type",
        "url": "",
      },
    ],
  ],
  "state": Object {},
}
`;

exports[`WebhooksIndexPage - Integration Test should flow: rendered 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "key": "WEBHOOK_EVENTS",
        "type": "get-some-type",
        "url": "",
      },
    ],
  ],
  "state": Object {},
}
`;
